extract_endpoint/src/extract_endpoint/azure_utils.py

```diff
@@ -17,16 +18,58 @@ class EnvVariables(enum.Enum):
     domain = 'EXTRACT_ENDPOINT_STORAGE_DOMAIN'
 
 
-def upload_bytes_to_azure(coords: StorageCoordinates, data: bytes, filename: str) -> bool:
-    azure_service = blob.BlockBlobService(account_name=coords.account,
-                                          account_key=coords.key,
-                                          custom_domain=coords.domain)
-    azure_service.create_blob_from_bytes(coords.container, filename, data)
-    return filename in [blob.name for blob in azure_service.list_blobs(coords.container)]
+class StorageProtocol(typing_extensions.Protocol):
+    def upload(self, data: bytes, filename: str) -> bool:
+        pass
 
+    def download(self, filename: str) -> bytes:
+        pass
 
-def download_bytes_from_azure(coords: StorageCoordinates, filename: str) -> str:
-    azure_service = blob.BlockBlobService(account_name=coords.account,
-                                          account_key=coords.key,
-                                          custom_domain=coords.domain)
-    return azure_service.get_blob_to_bytes(coords.container, filename).content
+
+class MockStorage:
+    def __init__(self) -> None:
+        self._upload_run_count = 0
+        self._download_run_count = 0
+
+    @property
+    def upload_run_count(self) -> int:
+        return self._upload_run_count
+
+    @property
+    def download_run_count(self) -> int:
+        return self._download_run_count
+
+    def upload(self, data: bytes, filename: str) -> bool:
+        if data and filename:
+            self._upload_run_count += 1
+        return True
+
+    def download(self, filename: str) -> bytes:
+        if filename:
+            self._download_run_count += 1
+        return filename.encode()
```

Comment:
I see what you're doing here. Here's an alternative:
- `upload` stores the data into a local dictionary using `filename` as the key
- `download` returns the data you put at the key of `filename`

That way setting up data to be downloaded is just `mock_storage.upload('filename', data)`

Reply:
👍 this is such a good call, thank you! Code & tests are much cleaner this way